sqlglot/optimizer/qualify_tables.py

```diff
@@ -96,5 +98,25 @@ def qualify_tables(
                 if isinstance(udtf, exp.Values) and not table_alias.columns:
                     for i, e in enumerate(udtf.expressions[0].expressions):
                         table_alias.append("columns", exp.to_identifier(f"_col_{i}"))
+                if isinstance(udtf, exp.Lateral) and dialect == "snowflake":
+                    _qualify_snowflake_lateral(udtf)
+
 
     return expression
+
+
+FLATTEN_COLUMNS = ["SEQ", "KEY", "PATH", "INDEX", "VALUE", "THIS"]
+
+
+def _qualify_snowflake_lateral(lateral: exp.Lateral) -> None:
+    if isinstance(lateral.this, exp.Anonymous) and lateral.this.this and lateral.this.this.upper() == "FLATTEN":
+        table_alias = lateral.args.get("alias")
+        if table_alias:
+            table_alias.set(
+                "columns",
+                [
+                    exp.Alias(this=lateral.this, alias=exp.to_identifier(col))
```

Comment:
The `lateral.this` part is probably repeating all the `FLATTEN`s, it doesn't seem right.